tfhe/src/core_crypto/experimental/algorithms/test/lwe_stair_keyswitch.rs

```diff
@@ -0,0 +1,296 @@
+use super::*;
+
+#[derive(Copy, Clone, Debug, PartialEq)]
+pub struct StairKSParam<Scalar: UnsignedInteger> {
+    pub log_precision: MessageModulusLog,
+    /// This value is unused but allows to identify the parameter optimization that was done
+    pub _log_mu: usize,
+    pub glwe_dimension: GlweDimension,
+    pub polynomial_size: PolynomialSize,
+    // phi = number of coefficients != 0 in a given key for partial keys
+    // this is the general phi for a GLWE secret key, which saturates to 2443 according to the
+    // paper
+    // phi == lwe_dimension + ks1_unshared_coeff_count + ks2_unshared_coeff_count
+    pub partial_glwe_secret_key_fill: PartialGlweSecretKeyRandomCoefCount,
+    pub bsk_glwe_noise_distribution: DynamicDistribution<Scalar>,
+    pub lwe_dimension: LweDimension,
+    pub ks1_lwe_noise_distribution: DynamicDistribution<Scalar>,
+    pub ks2_lwe_noise_distribution: DynamicDistribution<Scalar>,
+    pub pbs_level: DecompositionLevelCount,
+    pub pbs_base_log: DecompositionBaseLog,
+    pub ks1_level: DecompositionLevelCount,
+    pub ks1_base_log: DecompositionBaseLog,
+    pub ks2_level: DecompositionLevelCount,
+    pub ks2_base_log: DecompositionBaseLog,
+    /// The number of elements being dropped when going from the large key to the inter key
+    pub ks1_unshared_coeff_count: LweSecretKeyUnsharedCoefCount,
+    /// The number of elements being dropped when going from the inter key to the small key
+    pub ks2_unshared_coeff_count: LweSecretKeyUnsharedCoefCount,
+    pub ciphertext_modulus: CiphertextModulus<Scalar>,
+}
+
+pub const PRECISION_4_STAIR: StairKSParam<u64> = StairKSParam {
+    log_precision: MessageModulusLog(4),
+    _log_mu: 4,
+    glwe_dimension: GlweDimension(2),
+    polynomial_size: PolynomialSize(1024),
+    partial_glwe_secret_key_fill: PartialGlweSecretKeyRandomCoefCount(2048),
+    bsk_glwe_noise_distribution: DynamicDistribution::new_gaussian_from_std_dev(StandardDev(
+        3.16202663074765e-16,
+    )),
+    lwe_dimension: LweDimension(664),
+    ks1_lwe_noise_distribution: DynamicDistribution::new_gaussian_from_std_dev(StandardDev(
+        7.60713313301797e-9,
+    )),
+    ks2_lwe_noise_distribution: DynamicDistribution::new_gaussian_from_std_dev(StandardDev(
+        0.0000380960250519291,
+    )),
+    pbs_level: DecompositionLevelCount(1),
+    pbs_base_log: DecompositionBaseLog(22),
+    ks1_level: DecompositionLevelCount(1),
+    ks1_base_log: DecompositionBaseLog(13),
+    ks2_level: DecompositionLevelCount(6),
+    ks2_base_log: DecompositionBaseLog(2),
+    ks1_unshared_coeff_count: LweSecretKeyUnsharedCoefCount(922),
+    ks2_unshared_coeff_count: LweSecretKeyUnsharedCoefCount(462),
+    ciphertext_modulus: CiphertextModulus::new_native(),
+};
+
+fn lwe_encrypt_stair_keyswitch_pbs_decrypt_custom_mod<
```

Comment:
Function is a bit long
Could be split into smaller ones

Reply:
Test could be a only a stair KS with no PBS

Reply:
this was to test the atomic pattern, (though here the cleartext mul has been removed to avoid triggering the low error probability of the parameters), I'm going to keep it this way as it's closer to what's in the internal repo